lxcmap: removal global singleton map, and pass down pointer to map from daemon

ianvernon opened the pull request:
For future work around hiding `bpf.Map` around an interface for easier unit testing of the endpoint package, as well as code cleanliness, remove the global LXCMap variable from `pkg/lxcmap`. Create a new type, `LXCMap`, and create an instance of this in the `daemon`. Push this `lxcmap` into endpoints from the daemon side.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8795)
<!-- Reviewable:end -->

Commit:
lxcmap: rename LXCMap to GlobalLXCMap

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
lxcmap: factor LXCMap into a type

Maintain a globally accessible LXCMap for now. Have the former functions with no
receiver act on this global LXCMap. Implement functions with LXCMap receivers.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
refactor users of LXCMap to use an instance of it

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
lxcmap: remove global LXCMap

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on cilium/cmd/bpf_endpoint_delete.go:
```diff
@@ -37,8 +37,9 @@ var bpfEndpointDeleteCmd = &cobra.Command{
 		}
-
-		if err := lxcmap.DeleteEntry(ip); err != nil {
+		lxcMap := lxcmap.NewMap(lxcmap.MapName)
```
I don't think that this map is opened? (Similar for other CLI paths)

joestringer reviewed (changes requested):
Looks like a nice incremental step, only comment is the CLI usage doesn't look right.

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -224,7 +225,7 @@ func CleanupEndpoint() {
 // CleanupEndpoint() must be called before calling LaunchAsEndpoint() to ensure
 // cleanup of prior cilium-health endpoint instances.
-func LaunchAsEndpoint(baseCtx context.Context, owner regeneration.Owner, n *node.Node, mtuConfig mtu.Configuration) (*Client, error) {
+func LaunchAsEndpoint(baseCtx context.Context, owner regeneration.Owner, n *node.Node, mtuConfig mtu.Configuration, lxcMap *lxcmap.LXCMap) (*Client, error) {
```
Does this require to import `lxcmap` into this package or we can have an interface instead?

ianvernon replied:
I can hide this behind `Owner` for now and remove it as a member of `Endpoint`, which will remove this change entirely.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -281,6 +282,8 @@ type Endpoint struct {
 	// realizedBPFConfig is the config currently active in the BPF datapath.
 	realizedBPFConfig *bpfconfig.EndpointConfig
 
+	LXCMap *lxcmap.LXCMap `json:"-"`
```
TBH it feels odd to have a global variable "shared" across all entities. This could leave under daemon (which I believe it already does) and when the endpoint would like to access the LXCMap it would do it through the owner interface.

ianvernon replied:
Sure, I can do that for now.

Commit:
Fixup! cmd: do not close maps which were not opened

Commit:
Fixup! use LXCMap from Owner interface

Commit:
lxcmap: refactor `EndpointFrontend` interface

Instead of having the interface return the corresponding key and value types
used by pkg/maps/lxcmap have the interface require accessing the fields needed
to synthesize the BPF keys and values as needed in pkg/maps/lxcmap itself. This
removes the need to return lxcmap-specific types in the interface itself, making
implementers not have to be aware of `pkg/maps/lxcmap`.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
WIP

Commit:
f

Commit:
hmmm

Commit:
fix gofmt
